ocaml/xapi/xapi_vm_group_helpers.ml

```diff
@@ -0,0 +1,173 @@
+(*
+ * Copyright (C) 2024 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "xapi_vm_group_helpers" end)
+
+open D
+
+let check_breach ~__context ~group =
+  let vms =
+    Db.VM_group.get_VMs ~__context ~self:group
+    |> List.filter (fun vm ->
+           Db.VM.get_power_state ~__context ~self:vm = `Running
+       )
+  in
+  let get_host vm =
+    let host =
+      match Db.VM.get_resident_on ~__context ~self:vm with
+      | h when h <> Ref.null ->
+          Some h
+      | _ ->
+          None
+    in
+    host
+  in
+  match vms with
+  | [] | [_] ->
+      None
+  | vm :: _ ->
```

Comment:
This might not work if the `vm` happens to be a `None`, although a running VM must have a valid resident_on.

Reply:
I think the value here is represented the VM number of some group which is extract from `Db.VM_group.get_VMs`, it should be a list type. I presume it won`t be None

Reply:
And it could be more compact:
```
 let only_host_of_group ~__context ~group =
   Db.VM_group.get_VMs ~__context ~self:group
   |> List.filter_map (fun vm ->
       let vm_rec = Db.VM.get_record ~__context ~self:vm in
       match ( vm_rec.API.vM_power_state
             , vm_rec.API.vM_resident_on
       )
       with
       | `Running, h when h <> Ref.null ->
           Some h
       | _ ->
           None
   )
   |> function
   | [] | [_] ->
       None
   | h :: remaining -> (
       if List.exists (fun h' -> h' <> h) remaining then
         None
       else
         Some h
   )
```

Reply:
This line can be more compact:
```
if List.exists ((<>) h) remaining then
...

Reply:
> I think the value here is represented the VM number of some group which is extract from Db.VM_group.get_VMs, it should be a list type. I presume it won`t be None

I mean the `host` returned from `get_host`.